src/lib/cares_parse_ptr_reply.c

```diff
@@ -0,0 +1,188 @@
+  if (status != ARES_SUCCESS)
+    return status;
+  if (aptr + len + QFIXEDSZ > abuf + alen)
+    {
+      ares_free(ptrname);
+      return ARES_EBADRESP;
+    }
+  aptr += len + QFIXEDSZ;
+
+  /* Examine each answer resource record (RR) in turn. */
+  for (i = 0; i < (int)ancount; i++)
+    {
+      /* Decode the RR up to the data field. */
+      status = ares__expand_name_for_response(aptr, abuf, alen, &rr_name, &len);
+      if (status != ARES_SUCCESS)
+        break;
+      aptr += len;
+      if (aptr + RRFIXEDSZ > abuf + alen)
+        {
+          status = ARES_EBADRESP;
+          break;
+        }
+      rr_type = DNS_RR_TYPE(aptr);
+      rr_class = DNS_RR_CLASS(aptr);
+      rr_ttl = DNS_RR_TTL(aptr);
+      rr_len = DNS_RR_LEN(aptr);
+      aptr += RRFIXEDSZ;
+      if (aptr + rr_len > abuf + alen)
+        {
+          status = ARES_EBADRESP;
+          break;
+        }
+
+      if (rr_class == C_IN && rr_type == T_PTR)
+        {
+          if (strcasecmp(rr_name, ptrname) != 0)
+          {
+            /* question and answer don't match */
+            status = ARES_ENODATA;
+            break;
+          }
+          /* Allocate storage for this PTR answer appending it to the list */
+          ptr_curr = ares_malloc_data(ARES_DATATYPE_CPTR_REPLY);
+          if (!ptr_curr)
+            {
+              status = ARES_ENOMEM;
+              break;
+            }
+          if (ptr_last)
+            {
+              cares_ptr_reply_set_next(ptr_last, ptr_curr);
+            }
+          else
+            {
+              ptr_head = ptr_curr;
+            }
+          ptr_last = ptr_curr;
+
+          /* Decode the RR data and set hostname to it. */
+          status = ares__expand_name_for_response(aptr, abuf, alen, &ptr_host,
+                                                  &len);
+          if (status != ARES_SUCCESS)
+            {
+              break;
+            }
+
+          cares_ptr_reply_set_host(ptr_curr, ptr_host);
+          cares_ptr_reply_set_ttl(ptr_curr, rr_ttl);
+        }
+      else if (rr_class == C_IN && rr_type == T_CNAME)
+        {
+          /* Decode the RR data and replace ptrname with it. */
+          ares_free(ptrname);
+          ptrname = NULL;
+          status = ares__expand_name_for_response(aptr, abuf, alen, &ptrname,
+                                                  &len);
+          if (status != ARES_SUCCESS)
+            {
+              break;
+            }
+        }
+      else
+        {
+          /* wrong record type */
+          status = ARES_ENODATA;
+          break;
+        }
+
+      /* Don't lose memory in the next iteration */
+      ares_free(rr_name);
+      rr_name = NULL;
+
+      /* Move on to the next record */
+      aptr += rr_len;
+      if (aptr > abuf + alen)
+        {  /* LCOV_EXCL_START: already checked above */
+          status = ARES_EBADRESP;
+          break;
+        }  /* LCOV_EXCL_STOP */
```

Comment:
Some parse functions have this check after moving aptr, and some don't.  Should all of them have it?